Mathlib/RingTheory/PowerSeries/Inverse.lean

```diff
@@ -208,16 +216,79 @@ set_option linter.uppercaseLean3 false in
 set_option linter.uppercaseLean3 false in
 #align power_series.X_inv PowerSeries.X_inv
 
 @[simp]
-theorem smul_inv (r : k) (φ : PowerSeries k) : (r • φ)⁻¹ = r⁻¹ • φ⁻¹ :=
+theorem smul_inv (r : k) (φ : k⟦X⟧) : (r • φ)⁻¹ = r⁻¹ • φ⁻¹ :=
   MvPowerSeries.smul_inv _ _
 #align power_series.smul_inv PowerSeries.smul_inv
 
+/-- `firstUnitCoeff` is the non-zero coefficient whose index is `f.order`, seen as a unit of the
+  field. It is obtained using `divided_by_X_pow_order`, defined in `PowerSeries.Order`-/
+def firstUnitCoeff {f : k⟦X⟧} (hf : f ≠ 0) : kˣ := by
+  let d := f.order.get (order_finite_iff_ne_zero.mpr hf)
+  have f_const : coeff k d f ≠ 0 := by apply coeff_order
+  have : Invertible (constantCoeff k (divided_by_X_pow_order hf)) := by
+    apply invertibleOfNonzero
+    convert f_const using 1
+    rw [← coeff_zero_eq_constantCoeff, ← zero_add d]
+    convert (coeff_X_pow_mul (exists_eq_mul_right_of_dvd (X_pow_order_dvd
+      (order_finite_iff_ne_zero.mpr hf))).choose d 0).symm
+    exact (self_eq_X_pow_order_mul_divided_by_X_pow_order hf).symm
+  exact unitOfInvertible (constantCoeff k (divided_by_X_pow_order hf))
+
+/-- `Inv_divided_by_X_pow_order` is the inverse of the element obtained by diving a non-zero power
+series by the largest power of `X` dividing it. Useful to create a term of type `Units`, done in
+`Unit_divided_by_X_pow_order` -/
+def Inv_divided_by_X_pow_order {f : k⟦X⟧} (hf : f ≠ 0) : k⟦X⟧ :=
+  invOfUnit (divided_by_X_pow_order hf) (firstUnitCoeff hf)
+
+@[simp]
+theorem Inv_divided_by_X_pow_order_rightInv {f : k⟦X⟧} (hf : f ≠ 0) :
+    divided_by_X_pow_order hf * Inv_divided_by_X_pow_order hf = 1 :=
+  mul_invOfUnit (divided_by_X_pow_order hf) (firstUnitCoeff hf) rfl
+
+@[simp]
+theorem Inv_divided_by_X_pow_order_leftInv {f : k⟦X⟧} (hf : f ≠ 0) :
+    (Inv_divided_by_X_pow_order hf) * (divided_by_X_pow_order hf) = 1 := by
+  rw [mul_comm]
+  exact mul_invOfUnit (divided_by_X_pow_order hf) (firstUnitCoeff hf) rfl
+
+open scoped Classical
+
+/-- `Unit_of_divided_by_X_pow_order` is the unit power series obtained by dividing a non-zero
+power series by the largest power of `X` that divides it. -/
+def Unit_of_divided_by_X_pow_order (f : k⟦X⟧) : k⟦X⟧ˣ :=
+  if hf : f = 0 then 1
+  else
+    { val := divided_by_X_pow_order hf
+      inv := Inv_divided_by_X_pow_order hf
+      val_inv := Inv_divided_by_X_pow_order_rightInv hf
+      inv_val := Inv_divided_by_X_pow_order_leftInv hf }
+
+theorem isUnit_divided_by_X_pow_order {f : k⟦X⟧} (hf : f ≠ 0) :
+    IsUnit (divided_by_X_pow_order hf) :=
+  ⟨Unit_of_divided_by_X_pow_order f,
+    by simp only [Unit_of_divided_by_X_pow_order, dif_neg hf, Units.val_mk]⟩
+
+theorem Unit_of_divided_by_X_pow_order_nonzero {f : k⟦X⟧} (hf : f ≠ 0) :
+    ↑(Unit_of_divided_by_X_pow_order f) = divided_by_X_pow_order hf := by
+  simp only [Unit_of_divided_by_X_pow_order, dif_neg hf, Units.val_mk]
+
+theorem Unit_of_divided_by_X_pow_order_zero : Unit_of_divided_by_X_pow_order (0 : k⟦X⟧) = 1 := by
```

Comment:
Can this be `simp` or the linter complains?

Reply:
Nope, that's fine (and added)